pyro/infer/abstract_infer.py

```diff
@@ -297,11 +297,12 @@ def _traces(self, *args, **kwargs):
             self.posterior.run(*args, **kwargs)
         for _ in range(self.num_samples):
             model_trace = self.posterior()
-            replayed_trace = poutine.trace(poutine.replay(self.model, model_trace)).get_trace(*args, **kwargs)
-            yield (replayed_trace, 0., 0)
+            resampler = poutine.resample_posterior(self.model, model_trace)
+            resampled_trace = poutine.trace(resampler).get_trace(*args, **kwargs)
+            yield (resampled_trace, 0., 0)
 
     def marginal(self, sites=None):
         """
         Gets marginal distribution from posterior.
```

Comment:
@ahmadsalim Thanks for modifying the behaviour of this method! I personally find this change really helpful! I think we'd need to change the docs to "Gets marginal distribution of this predictive." (or something similar) too in case this change is acceptted. :)

Reply:
Sure, I will fix it!